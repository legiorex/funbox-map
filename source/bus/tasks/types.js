export const types = {
    // Sync
    FILL_TASKS:           'FILL_TASKS',
    CREATE_TASK:          'CREATE_TASK',
    REMOVE_TASK:          'REMOVE_TASK',
    TOGGLE_FAVORITE_TASK: 'TOGGLE_FAVORITE_TASK',
    EDIT_MESSAGE_TASK:    'EDIT_MESSAGE_TASK',

    // Async
    FETCH_TASKS_ASYNC: 'FETCH_TASKS_ASYNC',
    CREATE_TASK_ASYNC: 'CREATE_TASK_ASYNC',
    REMOVE_TASK_ASYNC: 'REMOVE_TASK_ASYNC',
    UPDATE_TASK_ASYNC: 'UPDATE_TASK_ASYNC',
};
