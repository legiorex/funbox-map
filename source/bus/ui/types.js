export const types = {
    START_FETCHING:      "START_FETCHING",
    STOP_FETCHING:       "STOP_FETCHING",
    EMIT_ERROR:          "EMIT_ERROR",
    INPUT_TASK:          "INPUT_TASK",
    CLEAR_TASK:          "CLEAR_TASK",
    CHECK_ALL_TASKS:     "CHECK_ALL_TASKS",
    START_EDITING_TASK:  "START_EDITING_TASK",
    FINISH_EDITING_TASK: "FINISH_EDITING_TASK",
    CURRENT_INPUT:       "CURRENT_INPUT",
    NEW_MESSAGE_TASK:    "NEW_MESSAGE_TASK",
    PREV_MESSAGE:        'PREV_MESSAGE',
};
