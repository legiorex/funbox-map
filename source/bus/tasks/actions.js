// Types
import { types } from "./types";

// Instruments
import { api } from '../../REST';

export const tasksActions = {

    // Sync

    fillTasks: (tasks) => {
        return {
            type:    types.FILL_TASKS,
            payload: tasks,
        };
    },
    createTask: (task) => {
        return {
            type:    types.CREATE_TASK,
            payload: task,
        };
    },

    // Async

    fetchTasksAsync: () => {
        return {
            type: types.FETCH_TASKS_ASYNC,
        };

    },
    createTaskAsync: (newTaskMessage) => {
        return {
            type:    types.CREATE_TASK_ASYNC,
            payload: newTaskMessage,
        };
    },

};
