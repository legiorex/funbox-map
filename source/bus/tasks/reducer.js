// Core
import { fromJS, List } from 'immutable';

// Instruments
import { types } from "./types";

const initialState = List();

export const tasksReducer = (state = initialState, action) => {

    switch (action.type) {
        case types.FILL_TASKS:
            return fromJS(action.payload);

        case types.CREATE_TASK:
            return state.unshift(fromJS(action.payload));
        default:
            return state;
    }
};
