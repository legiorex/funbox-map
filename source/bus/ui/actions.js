// Types
import { types } from './types';

export const uiActions = {
    startFetching: () => {
        return {
            type: types.START_FETCHING,
        };
    },

    stopFetching: () => {
        return {
            type: types.STOP_FETCHING,
        };
    },
    inputTask: (value) => {

        return {
            type:    types.INPUT_TASK,
            payload: value,
        };
    },
    clearTask: () => {
        return {
            type: types.CLEAR_TASK,
        };
    },

    checkedAllTasks: (tasks) => {

        return {
            type:    types.CHECK_ALL_TASKS,
            payload: tasks,
        };
    },

    editTaskMessage: (newMessage) => {
        return {
            type:    types.NEW_MESSAGE_TASK,
            payload: newMessage,
        };
    },
    
    startEditingTask: (idTask) => {
        return {
            type:    types.START_EDITING_TASK,
            payload: idTask,
        };
    },
    finishEditingTask: () => {
        return {
            type: types.FINISH_EDITING_TASK,
        };
    },

    prevMessage: (message) => {
        return {
            type:    types.PREV_MESSAGE,
            payload: message,
        };
    },

    emitError: (error, meta = null) => {
        return {
            type:    types.EMIT_ERROR,
            payload: error,
            error:   true,
            meta,
        };
    },
};
