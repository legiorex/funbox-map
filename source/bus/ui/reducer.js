// Core
import { Map } from 'immutable';

// Types
import { types } from './types';

const initialState = Map({
    checkedAllTasksCompleted: false,
    isFetching:               false,
    valueInputTask:           '',
    newMessage:               '',
    prevMessage:              '',
    editingTaskId:            null, //или нул или строка id

});

export const uiReducer = (state = initialState, action) => {
    switch (action.type) {
        case types.START_FETCHING:
            return state.set('isFetching', true);

        case types.STOP_FETCHING:
            return state.set('isFetching', false);

        case types.INPUT_TASK:
            return state.set('valueInputTask', action.payload);

        case types.CLEAR_TASK:
            return state.set('valueInputTask', '');

        case types.CHECK_ALL_TASKS:

            const checked = action.payload.every((task) => {
                return task.get('completed');
            });

            return state.set('checkedAllTasksCompleted', checked);

        case types.START_EDITING_TASK:
            return state.set('editingTaskId', action.payload);

        case types.FINISH_EDITING_TASK:
            return state.set('editingTaskId', null);

        case types.NEW_MESSAGE_TASK:
            return state.set('newMessage', action.payload);

        case types.PREV_MESSAGE:
            return state.set('prevMessage', action.payload);

        default:
            return state;
    }
};
