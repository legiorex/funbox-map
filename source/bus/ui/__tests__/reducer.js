// Core
import { fromJS, Map } from "immutable";

// Instruments
import { types } from '../types';
import { uiActions } from "../actions";
import { uiReducer } from "../reducer";
import { sortTasksByGroup } from '../../../instruments/helpers';

const initialState = Map({
    checkedAllTasksCompleted: false,
    isFetching:               false,
    valueInputTask:           '',
    newMessage:               '',
    prevMessage:              '',
    editingTaskId:            null, //или нул или строка id

});

describe('ui reducer', () => {
    test('should return initial state by default', () => {
        expect(uiReducer(void 0, {})).toEqual(initialState);
    });
    test('should handler START_FETCHING action', () => {
        expect(uiReducer(void 0, uiActions.startFetching())).toEqual(
            initialState.set("isFetching", true)
        );
    });

    test('should handler STOP_FETCHING action', () => {
        expect(uiReducer(void 0, uiActions.stopFetching())).toEqual(
            initialState.set('isFetching', false)
        );
    });

    test('should handler INPUT_TASK action', () => {
        expect(uiReducer(void 0, uiActions.inputTask(__.message))).toEqual(
            initialState.set('valueInputTask', __.message)
        );
    });

    test('should handler CLEAR_TASK action', () => {
        expect(
            uiReducer(void 0, uiActions.clearTask())
        ).toEqual(initialState.set('valueInputTask', ''));
    });

    test('should handler CHECK_ALL_TASKS action', () => {
        expect(
            uiReducer(void 0, uiActions.checkedAllTasks(__.tasks))
        ).toEqual(initialState.set('checkedAllTasksCompleted', __.tasks.every((task) => {
            return task.get('completed');
        })));
    });

    test('should handler START_EDITING_TASK action', () => {
        expect(
            uiReducer(void 0, uiActions.startEditingTask(__.taskID))
        ).toEqual(initialState.set('editingTaskId', __.taskID));
    });

    test('should handler FINISH_EDITING_TASK action', () => {
        expect(
            uiReducer(void 0, uiActions.finishEditingTask())
        ).toEqual(initialState.set('editingTaskId', null));
    });

    test('should handler NEW_MESSAGE_TASK action', () => {
        expect(
            uiReducer(void 0, uiActions.editTaskMessage(__.message))
        ).toEqual(initialState.set('newMessage', __.message));
    });

    test('should handler PREV_MESSAGE action', () => {
        expect(
            uiReducer(void 0, uiActions.prevMessage(__.message))
        ).toEqual(initialState.set('prevMessage', __.message));
    });

});
