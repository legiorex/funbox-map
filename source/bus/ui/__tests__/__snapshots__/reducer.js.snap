// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ui reducer should handler CHECK_ALL_TASKS action 1`] = `
Object {
  "payload": Immutable.List [
    Array [
      "id",
      "TEST_ID",
    ],
    Array [
      "message",
      "TEST_MESSAGE",
    ],
    Array [
      "completed",
      true,
    ],
    Array [
      "favorite",
      false,
    ],
    Array [
      "created",
      "TEST_CREATED",
    ],
    Array [
      "modified",
      "TEST_MODIFIED",
    ],
  ],
  "type": "CHECK_ALL_TASKS",
}
`;
